From 251911de9d4116c385d173bbff033f4023002b0f Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Tue, 04 Oct 2022 08:50:39 +0300
Subject: [PATCH] [SW-69872] gaudi3: add delay post slave detection

Currently for inter die notification, master holds a pulse of 10us
and slave detects the same. But logically, slave can detect the pulse
at any point on the 0-10th microsecond. Due to this, if the pulse is
detected on slave in the first half pulse, slave starts executing
whilst master is still completing the indication pulse.

Due to this there is a possibility that slave mis-interprets this pulse
when checked again in the execution flow.

To avoid such race, a delay is added on slave so that each pulse is
detected only once on slave.

Change-Id: I0a055de83c42042edc5d32e6409c7e732e022c6a
---

diff --git a/arch/gaudi3/arc/slave.c b/arch/gaudi3/arc/slave.c
index 5e56164..3713f3c 100644
--- a/arch/gaudi3/arc/slave.c
+++ b/arch/gaudi3/arc/slave.c
@@ -31,6 +31,12 @@
 	while (1) {
 		val = regmap_readl(mmD0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN);
 		if (val & D0_PSOC_GLOBAL_CONF_DIE_RECOVERY_IN_IND_M) {
+			/* Master die holds the indication pulse for 10us,
+			 * Hence we need to wait for 10us after detection
+			 * to make sure slave follows the master in terms
+			 * of point of execution
+			 */
+			udelay(NOTIFY_SLV_DURATION_US);
 			break;
 		}
 	}
